torchtune/models/flamingo/_encoder.py

```diff
@@ -0,0 +1,125 @@
+from torchtune.modules.transformer import _get_clones
+
+
+class FlamingoProjectionHead(nn.Module):
+    """Projection transformer to adapt the output of a
+    pretrained encoder to a pretrained decoder model.
+
+    Args:
+        layer (nn.Module): Transformer Decoder layer
+        num_layers (int): Number of Transformer Decoder layers
+        output (nn.Module): Output linear layer. Input dim is
+            (num_hidden + 1) * encoder_dim and output is decoder_dim.
+        num_hidden_inputs (int): Number of expected hidden state inputs
+    """
+
+    def __init__(
+        self,
+        layer: nn.Module,
+        num_layers: int,
+        output: nn.Module,
+        num_hidden_inputs: int = 0,
+    ) -> None:
+        super().__init__()
+        self.layers = _get_clones(layer, num_layers)
+        self.output = output
+        self.num_hidden = num_hidden_inputs
+
+    def forward(
+        self,
+        x: Tensor,
+        hidden_states: Optional[List[Tensor]] = None,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape [b x i x t x e x d]
+            hidden_states (Optional[List[Tensor]]): list of hidden states
+                from the encoder. Each hidden state has the same shape as x.
+
+        Returns:
+            Tensor: output tensor of a sequence of embedings [b x s x d]
+                where sequence length is num_imgs*num_tiles+num_embeds
+
+        Notation used for tensor shapes:
+            - b: batch size
+            - i: number of images
+            - t: number of tiles (where a single image is broken into multiple tiles)
+            - e: number of embeds per tile (e.g. CLS embed + patch embeds, etc.)
+            - s: sequence length computed by i*t*e
+            - d: embed dim
+        """
+        bsz, imgs, tiles, embeds, dim = x.shape
+
+        # apply transformer layers
+        x = x.view(bsz * imgs, tiles * embeds, dim)
+        for layers in self.layers:
+            x = layers(x)
+        x = x.view(bsz, imgs, tiles, embeds, dim)
+
+        # interleave hidden states and cat with x
+        if self.num_hidden > 0:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.view(bsz, imgs, tiles, embeds, -1)
+            x = torch.cat([x, hidden_states], dim=-1)
+
+        # shape [b x s x d]
+        x = self.output(x).reshape(bsz, imgs * tiles * embeds, -1)
+
+        return x
+
+
+class FlamingoEncoder(nn.Module):
+    """Vision encoder model for Flamingo. This combines a pretrained
+    vision encoder with a learnable projection head. The projection head
+    is converted to a fusion module and supports fusion utils.
+
+    Args:
+        clip (nn.Module): CLIP encoder vision model
+        projection_head (nn.Module): projection_head that takes embeddings
+            with dimension encoder_dim as input and outputs embeddings of
+            size decoder_dim.
```

Comment:
nbd to leave these both as general nn.Modules but they do have very specific signatures that make it hard to plug in any old nn.Module. Maybe point to the relevant CLIP and projection classes as an example or something

Reply:
This being inside of the flamingo folder it's not really meant to be reused for other purposes. I can just set the type to be the specific modules.